Allow manual `xhr.send(...)` when the `upload-request` event is prevented

platosha opened the pull request:
Resolves #96 without the API changes.

When the `upload-request` event is prevented, the `xhr.send(formData)` is not called in the upload. At the same time, a couple of state file properties are not updated in this case. Therefore, it's hard to send the xhr manually.

This PR allows developers to call `xhr.send(whatever)` manually in the `upload-request` event listener, while preventing the default send call. This also makes it possible to send files directly without a FormData wrapper, as well as allows custom data wrappers/encodings.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-upload/113)
<!-- Reviewable:end -->

manolo commented:
Apart from this I would like to have a declarative way to send the file as a blob. 
IMO, It's a very common feature and making the user utilise javascript to handle any file to upload is not the best UX.

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[vaadin-upload.html, line 677 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-KKNAVeRtuxh5ruoDIHq:-KKNAVeRtuxh5ruoDIHr:276415586) ([raw file](https://github.com/vaadin/vaadin-upload/blob/d9d2e8a79146b89635c0314fb114fdbe53212a4d/vaadin-upload.html#L677)):*
> ```HTML
>       }
> 
>       this.fire('upload-start', {file: file, xhr: xhr});
> ```

I'm not sure about  continuing with the workflow after evt is prevented.

To me makes more sense this:

```
file.complete = ...
this._nofifyFileChanges(file) 
this.fire ('upload-start', ...)

evt = this.fire('upload-request', ...
if (!evt.defaultPrevented) ...

```

---

*[demo/advanced.html, line 85 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-KKNBpleeG1ZBwz845iF:-KKNBplfnLB52XZhmKg2:-1374467023) ([raw file](https://github.com/vaadin/vaadin-upload/blob/d9d2e8a79146b89635c0314fb114fdbe53212a4d/demo/advanced.html#L85)):*
> ```HTML
>             // Send the file without the FormData wrapper:
>             // event.preventDefault();
>             // xhr.send(event.detail.file);
> ```

I would add a new example instead of having these comments here which could be confusing

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-:-KKNC1kw4ZRI_Cc54Aaa:-1061594991)*
<!-- Sent from Reviewable.io -->

platosha commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[vaadin-upload.html, line 677 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-KKNAVeRtuxh5ruoDIHq:-KKNFT9Zm8LwemxOeIn1:-1527806559) ([raw file](https://github.com/vaadin/vaadin-upload/blob/d9d2e8a79146b89635c0314fb114fdbe53212a4d/vaadin-upload.html#L677)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary>
> I'm not sure about  continuing with the workflow after evt is prevented.
> 
> To me makes more sense this:
> 
> ```
> file.complete = ...
> this._nofifyFileChanges(file) 
> this.fire ('upload-start', ...)
> 
> evt = this.fire('upload-request', ...
> if (!evt.defaultPrevented) ...
> 
> ```
> 
</details>
In your example there're two consecutive events, which doesn't make a lot sense.

Also, `'upload-start'` by doc is fired after the request is sent.

IMO, not interrupting the flow after preventing `upload-request` is OK. At the point when `upload-request` is fired, the uploading request is already open, so the uploading has already started, the connection has opened. The only thing you can prevent here is the default request send, you can’t cancel the uploading overall.

---

*[demo/advanced.html, line 85 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-KKNBpleeG1ZBwz845iF:-KKNLGaT3l04MR5EeTz1:-499146009) ([raw file](https://github.com/vaadin/vaadin-upload/blob/d9d2e8a79146b89635c0314fb114fdbe53212a4d/demo/advanced.html#L85)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary>
> I would add a new example instead of having these comments here which could be confusing
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[vaadin-upload.html, line 677 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-KKNAVeRtuxh5ruoDIHq:-KKhb-zA-kwnukpiGAaL:-323065025) ([raw file](https://github.com/vaadin/vaadin-upload/blob/d9d2e8a79146b89635c0314fb114fdbe53212a4d/vaadin-upload.html#L677)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary>
> In your example there're two consecutive events, which doesn't make a lot sense.
> 
> Also, `'upload-start'` by doc is fired after the request is sent.
> 
> IMO, not interrupting the flow after preventing `upload-request` is OK. At the point when `upload-request` is fired, the uploading request is already open, so the uploading has already started, the connection has opened. The only thing you can prevent here is the default request send, you can’t cancel the uploading overall.
</details>
Preventing the default makes it possible to trigger the send asynchronously, right? Then `'upload-start'` would get fired before the request is actually sent.

Would make more sense to stop the workflow on preventDefault in that case.

---

*[docs/vaadin-upload-basic.adoc, line 113 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-KKhZ5zZnY8rWjkC2EcR:-KKhZ5zZnY8rWjkC2EcS:-360068415) ([raw file](https://github.com/vaadin/vaadin-upload/blob/2bc2095df5f1e1c23eeaaa3312367485ada60f67/docs/vaadin-upload-basic.adoc#L113)):*
> ```AsciiDoc
> `upload-progress`:: Fired as many times as a file progress is updated.
> `upload-request`:: Fired when the request has been opened but not yet sent. It is useful for appending additional data keys to the request and for changing some parameters such as headers.
>   If the event is default-prevented, then the request is not sent to the server. In this case, you can send XHR manually.
> ```

...send the (related?) XHR...

---

*[docs/vaadin-upload-server.adoc, line 68 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-KKhZGBzvjzx6u1lNnpv:-KKhZGBzvjzx6u1lNnpw:1440347766) ([raw file](https://github.com/vaadin/vaadin-upload/blob/2bc2095df5f1e1c23eeaaa3312367485ada60f67/docs/vaadin-upload-server.adoc#L68)):*
> ```AsciiDoc
> ----
> 
> `upload-request`:: You can change request parameters before it is sent, append data keys to the request.
> ```

...sent or append...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113)*
<!-- Sent from Reviewable.io -->

platosha commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-upload.html, line 677 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-KKNAVeRtuxh5ruoDIHq:-KKhv-gFaPQJ_pqVQ4rm:-1653327959) ([raw file](https://github.com/vaadin/vaadin-upload/blob/d9d2e8a79146b89635c0314fb114fdbe53212a4d/vaadin-upload.html#L677)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> Preventing the default makes it possible to trigger the send asynchronously, right? Then `'upload-start'` would get fired before the request is actually sent.
> 
> Would make more sense to stop the workflow on preventDefault in that case.
</details>
Right. Changed the workflow with the async send in mind.

---

*[docs/vaadin-upload-basic.adoc, line 113 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-KKhZ5zZnY8rWjkC2EcR:-KKhvHmdI8U6j5TpIO6c:-499146009) ([raw file](https://github.com/vaadin/vaadin-upload/blob/2bc2095df5f1e1c23eeaaa3312367485ada60f67/docs/vaadin-upload-basic.adoc#L113)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> ...send the (related?) XHR...
</details>
Done.

---

*[docs/vaadin-upload-server.adoc, line 68 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-KKhZGBzvjzx6u1lNnpv:-KKhvKqai_LRdW-1TKom:-849436948) ([raw file](https://github.com/vaadin/vaadin-upload/blob/2bc2095df5f1e1c23eeaaa3312367485ada60f67/docs/vaadin-upload-server.adoc#L68)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> ...sent or append...
</details>
Changed to `and`, I think it’s more appropriate in the possibilities list.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113)*
<!-- Sent from Reviewable.io -->

manolo commented:
Reviewed 1 of 5 files at r1, 1 of 1 files at r2, 4 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 5 files at r1, 1 of 1 files at r2, 4 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-:-KKsPJxfe7qM45-Zi-1C:1417088721)*
<!-- Sent from Reviewable.io -->

tehapo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/113#-:-KLGQKj8i_1v2C9CZxfh:1417088721)*
<!-- Sent from Reviewable.io -->
